### ISSUE TITLE
Problems when declaring rec-time independent layers within rec-layer

### ISSUE THREAD

**Issue by Zettelkasten** [rwth-i6/returnn#500]
I am using a basic cross-attention setup inside a rec-layer, essentially as in e.g. [`test_reclayer_optimize_out_dot` here](https://github.com/rwth-i6/returnn/blob/522ce85122103fc8a21df5798e9596476250a96d/tests/test_TFNetworkRecLayer.py#L3435).
In that test case, the layers for calculating keys and value from the encoder states are declared outside the rec layer, as these are completely independent from the rec-layer time.
However, in my case I would like to declare them *within* the rec layer, right next to the rest of the attention calculation.
(That would simplify my code as everything related to the attention would be in one place, and I don't think there really is a reason this should not be supported. I use `optimize_move_layers_out=True` always anyway.)

I added a test case in PR #501:
```python3
  check_reclayer_optimize_out(
    {"class": "linear", "activation": None, "from": ["att"]},
    other_subnet_layers={
      "s": {"class": "linear", "activation": None, "with_bias": False, "from": ["data:source"],
            "n_out": EncKeyTotalDim},  # (B, D)  -- Q (query). D should be same as enc_ctx
      "att_query": {"class": "split_dims", "axis": "F", "dims": (AttNumHeads, EncKeyPerHeadDim),
                    "from": ["s"]},  # (B, H, D/H)
      # this does not depend on the rec-time, but you should still be able to define it here.
      "enc_ctx0": {"class": "linear", "activation": None, "with_bias": False, "from": ["base:encoder"],
                   "n_out": EncKeyTotalDim},  # (B, enc-T, D)
      "enc_ctx": {"class": "split_dims", "axis": "F", "dims": (AttNumHeads, EncKeyPerHeadDim),
                  "from": ["enc_ctx0"], "is_output_layer": True},  # (B, enc-T, H, D/H)
      "enc_value0": {"class": "linear", "activation": None, "with_bias": False, "from": ["base:encoder"],
                     "n_out": EncValueTotalDim},  # (B, enc-T, D)
      "enc_value": {"class": "split_dims", "axis": "F", "dims": (AttNumHeads, EncValuePerHeadDim),
                    "from": ["enc_value0"], "is_output_layer": True},  # (B, enc-T, H, D/H)
      "energy": {"class": "dot", "red1": -1, "red2": -1, "var1": "T", "var2": "T?",  # Note the "T?".
                 "from": ["enc_ctx", "att_query"]},
      "att_weights": {"class": "softmax_over_spatial", "from": ["energy"]},  # (B, enc-T, H, 1)
      "att0": {"class": "generic_attention", "weights": "att_weights", "base": "enc_value"},  # (B, H, V)
      "att": {"class": "merge_dims", "axes": "static", "from": ["att0"]},  # (B, H*V); Use "static" here.
      },
    shared_base_net={
      "encoder": {"class": "copy", "from": ["data"]}
    },
    rtol=1e-3)
```
(note that `enc_ctx0`, `enc_value0`, etc. are declared within the subnet, not in the base net.)

There is an error here, and the `encoder` layer of the base never gets constructed, even though the subnet of course depends on it (see the [stack trace in the PR here](https://github.com/rwth-i6/returnn/runs/2478555137?check_suite_focus=true)).

I also noted that if there is another layer in the subnet which depends on `encoder` (for example if I calculate the key in the base net, and the values in the rec subnet), then it works.

Can someone point me to where in the subnet construction code this fails so I can try to fix it?
Thanks!

**Comment by albertz** [comment]
Hm strange, I'm pretty sure I did exactly this and it worked already.

However, on the test case, there is one bug: You need to set `"is_output_layer": True` on the `encoder` layer for this `shared_base_net`, otherwise it will not get constructed. (I will comment on the PR specifically for this.)

Does this also happen in some real case, or when you fix the PR?

**Comment by Zettelkasten** [comment]
> However, on the test case, there is one bug: You need to set `"is_output_layer": True` on the `encoder` layer for this `shared_base_net`, otherwise it will not get constructed. (I will comment on the PR specifically for this.)

Oh, that does fix the test case in the PR. I updated it .. but I wonder, why is it necessary? Shouldn't it work without, too?

> Does this also happen in some real case, or when you fix the PR?

My real case is super similar, but in contrary to the one I pushed still fails with this change - however with a different error. Now some shapes don't match. I simplified it and added that too ([see `test_reclayer_att_with_kv_in_rec`](https://github.com/rwth-i6/returnn/pull/501/commits/98294eeb990421eb5c7f7b3613d781b5d35cf086)), with error log [here](https://github.com/rwth-i6/returnn/runs/2479307005?check_suite_focus=true).
Again here the code works fine if `att_key0` and `att_value0` etc. are moved outside the rec layer.

**Comment by albertz** [comment]
> Oh, that does fix the test case in the PR. I updated it .. but I wonder, why is it necessary? Shouldn't it work without, too?

No. See the logic of this test, how it handles `shared_base_net`. This is really only specific about `shared_base_net` here.

In your new test case (`test_reclayer_att_with_kv_in_rec`), or in general, you do not need `is_output_layer`.
In your new test case, the error
```
ValueError: Inconsistent shapes: saw (?, 2, 3) but expected (?, ?, 2, 3) 
```
occurs for `layer = <local> <SplitDimsLayer output/'att_value' out_type=Data(batch_shape_meta=[B,T|'time:var:extern_data:data',2,F|3])>`.
There is some logic which checks if an optimized layer should be unrolled when used inside the network. This is usually the case, when it uses the same frame axis as the rec layer. But in your examples, it is not the case, as it is totally independent. This is the check: https://github.com/rwth-i6/returnn/blob/522ce85122103fc8a21df5798e9596476250a96d/returnn/tf/layers/rec.py#L2180
It seems this logic fails for your new test case. I assume because the time dim tag is unknown maybe at the time of the check. Or so. Can you maybe set a breakpoint there and check?

**Comment by Zettelkasten** [comment]
> No. See the logic of this test, how it handles `shared_base_net`. This is really only specific about `shared_base_net` here.

Ahh, I understand. Cool thanks! :D

> It seems this logic fails for your new test case. I assume because the time dim tag is unknown maybe at the time of the check. Or so. Can you maybe set a breakpoint there and check?

Here, breaking right before the line you linked:

```
-> if not layer.output.mark_same_time(rec_layer.output):
(Pdb) layer
<SplitDimsLayer output/'att_value' out_type=Data(batch_shape_meta=[B,T|'time:var:extern_data:data',2,F|3])>
(Pdb) layer.output
Data(name='att_value_output', batch_shape_meta=[B,T|'time:var:extern_data:data',2,F|3])
(Pdb) rec_layer.output
Data(name='output_output', dtype='int32', sparse=True, dim=6, batch_shape_meta=[T|'time:var:extern_data:data',B&Beam{'output/output'}(4)])
```

So - both dim tags exist, but `rec_layer` also has the `data` and not the `classes` time dim. Is that the issue?

**Comment by albertz** [comment]
> ```
> -> if not layer.output.mark_same_time(rec_layer.output):
> (Pdb) layer
> <SplitDimsLayer output/'att_value' out_type=Data(batch_shape_meta=[B,T|'time:var:extern_data:data',2,F|3])>
> (Pdb) layer.output
> Data(name='att_value_output', batch_shape_meta=[B,T|'time:var:extern_data:data',2,F|3])
> (Pdb) rec_layer.output
> Data(name='output_output', dtype='int32', sparse=True, dim=6, batch_shape_meta=[T|'time:var:extern_data:data',B&Beam{'output/output'}(4)])
> ```
> 
> So - both dim tags exist, but `rec_layer` also has the `data` and not the `classes` time dim. Is that the issue?

Oh, yes that looks incorrect. `rec_layer.output` looks incorrect. It should have the shape `[T|'time:var:extern_data:classes',B&Beam{'output/output'}(4)]`.

Maybe because you have `search_flag=True, train_flag=False, eval_flag=False`, it will ignore the target? Actually not sure if it would make sense to use `classes` in this case, as it actually should not use this extern data key. But then the rec layer should get its own dim tag. This might be related to the dim tag issues we discussed in #391.

**Comment by Zettelkasten** [comment]
I noticed that also when `search_flag=False, train_flag=True, eval_flag=False`, that in `_SubnetworkRecCell.get_output`, the `rec_layer` also is

    <RecLayer 'output' out_type=Data(dtype='int32', sparse=True, dim=6, batch_shape_meta=[T|'time:var:extern_data:data',B])>

But then, after the entire network is constructed, it's (correctly) set to `<RecLayer 'output' out_type=Data(dtype='int32', sparse=True, dim=6, batch_shape_meta=[T|'time:var:extern_data:classes',B])>`.
Is that correct this way? Which code is responsible for actually setting the time dim - and is it set correctly *after* `get_output` is called?

**Comment by albertz** [comment]
> Is that correct this way?

Well, I suppose there were some reasons, or maybe this was a workaround, or maybe this did not turn out as a problem so far.

I'm not really sure why/how it gets `time:var:extern_data:data` initially. This would be interesting to see, where this comes from.

In any case, obviously this is not correct. It should never be unexpected. The expected and correct thing would be to determine in advance the correct dim tag, and set it in advance. It either can be:

- If it has some input, this defines the frame axis, and dim tag.
- Some new dim tag. (E.g. in search.)
- When the targets are used inside for some framewise loss (for eval or training without search), this defines it.

I guess it was a bit tricky to always correctly determine this in advance.

We could also create a new dim tag in any case (or in any of the non-obvious cases) and it can later be defined as being equal to another existing dim tag (`DimensionTag.same_as`).

**Comment by albertz** [comment]
I started some initial implementation in #527.

**Comment by Zettelkasten** [comment]
Just before you added #527, I found the reason why the time dim was set to `data` instead of `classes`:
In `get_out_data_from_opts` of `RecLayer`, the output `size_placeholder` is set by the generic `_post_init_output` implementation of `LayerBase`.
In case there are `sources` to the layer, that function will copy the time dim from them. In my case, I did not supply any `sources`, so RETURNN used the `data` default - thus also falsely taking the `data` time dim as output for my rec layer.
Instead, when setting `'from': []`, `_post_init_output` will not set the time dim to `data`. As far as I can tell, it is then left unset or set to the target time dim during training.
 
So for me, changing `RecLayer` s.t. the default `from` is always `[]` whenever there is a subnet fixes this for now.
And in total, this is also a very good reason to remove the `sources` default to use data entirely (#519) - we had this problem for three weeks (at least closely related) to this default! :D

But still - the real issue remains in the logic of how `RecLayer` determines it's output time dim. E.g. this will fail if one really wants to use `data` as input to the RecLayer (I never use any inputs and just access them using `base:` - but that's supported still).
So that would be great to change in #527.

We can also change the logic in `_post_init_output`. E.g. we could always use the `target` time axis to create a missing size placeholder, and prioritize that over the `sources` time axis. But it's just better to always set the time axis explicitly instead of relying on such heuristics that will fail eventually.

**Comment by Zettelkasten** [comment]
Hm, some tests actually specify a rec subnet without giving any explicit sources and depend on it being `data`, e.g. `test_rec_subnet_simple_rnn`:
```python3
"network": {
  "output": {
    "class": "rec",
    "unit": {
      # Recurrent subnet here, operate on a single time-step:
      "output": {
        "class": "linear",
        "from": ["prev:output", "data:source"],
        "activation": "relu",
        "n_out": n_out},
    },
    "n_out": n_out},
}
```

I think I will just go ahead and change these tests to explicitly specify `from`, we anyway want to do that eventually for #519.

For those tests the time rec layer output time axis will then be set wrongly, but at least it's identical to the behavior before.

**Comment by albertz** [comment]
Yes, #527 is about setting the time-dim in advance, for all purpose where it might be needed. E.g. https://github.com/rwth-i6/returnn/blob/522ce85122103fc8a21df5798e9596476250a96d/returnn/tf/layers/rec.py#L2180 (I thought this is the problem in this issue here but it seems not) but also other cases (e.g. related to #391).

Ok, so to clarify, your problem was that you had `from` set incorrectly? Then it works? (I ask because #501 still seems to fail.) So then we can probably close this issue here (after you fix the test case in #501).

Yes, I fully agree that the default `from` (in general, not just for RecLayer) is very annoying. I got hit by this a couple of times as well. But this is #519.

**Comment by Zettelkasten** [comment]
Okay yes! That's good.

#501 is now only failing because of other tests, because I set the default. But as you said, let's leave those changes for #519, and then there we will also fix the other cases that rely on the default. Once the tests run through and pass, I'll close this issue.
